lib/newrelic_riak/riak_client.rb

```diff
@@ -8,104 +9,80 @@
-      add_method_tracer :set_bucket_props, 'Database/Riak/set_bucket_props'
-
-      add_method_tracer :mapred, 'Database/Riak/mapred'
-
-      add_method_tracer :list_keys, 'Database/Riak/list_keys'
-      add_method_tracer :reload_object, 'Database/Riak/reload_object'
-      add_method_tracer :delete_object, 'Database/Riak/delete_object'
+    # Instrument the ProtoBuffsBackend
+    NewRelic::Agent.logger.info 'Riak: BeefcakeProtoBuffsBackend'
+    ::Riak::Client::BeefcakeProtobuffsBackend.class_eval do
+      method_list = %w(ping list_buckets get_bucket_props set_bucket_props mapred list_keys
+                       reload_object delete_object server_info get_client_id set_client_id get_index)
+      method_list.each do |method_name|
+        NewRelic::Agent::Datastores.trace self, method_name, 'Riak'
+      end
+    end
+    
+    # Instrument RObject
+    NewRelic::Agent.logger.info 'Riak: RObject'
+    ::Riak::RObject.class_eval do
+      NewRelic::Agent::Datastores.trace self, 'serialize', 'Riak'
     end
 
+    # Instrument Riak client get/store
+    NewRelic::Agent.logger.info 'Riak: Wrap Riak::Client save/find with NR and alias them'
     ::Riak::Client.class_eval do
+      include NewRelic::Agent::Instrumentation::Riak
+      alias_method :store_object_without_newrelic, :store_object
+      alias_method :store_object, :store_object_with_newrelic
+      alias_method :get_object_without_newrelic, :get_object
+      alias_method :get_object, :get_object_with_newrelic
+    end
+  end
+end
 
-      def store_object_with_newrelic_trace(*args, &blk)
-        if NewRelic::Agent::Instrumentation::MetricFrame.recording_web_transaction?
-          total_metric = 'ActiveRecord/allWeb'
-        else
-          total_metric = 'ActiveRecord/allOther'
-        end
-
-        robject = args[0].is_a?(Array) ? args[0][0] : args[0]
-        bucket = robject.respond_to?(:bucket) && robject.bucket ? robject.bucket.name : ''
-        bucket = self.newrelic_riak_camelize(bucket)
-        metrics = ["ActiveRecord/#{bucket}/save", total_metric, 'ActiveRecord/#{bucket}', 'ActiveRecord/save', 'ActiveRecord/all']
-
-        self.class.trace_execution_scoped(metrics) do
-          start = Time.now
-
-          begin
-            store_object_without_newrelic_trace(*args, &blk)
-          ensure
-            s = NewRelic::Agent.instance.transaction_sampler
-            s.notice_nosql(args.inspect, (Time.now - start).to_f) rescue nil
+module NewRelic
+  module Agent
+    module Instrumentation
+      module Riak
+        def get_set_callback
+          Proc.new do |result, scoped_metric, elapsed|
+            # See datastores.rb in newrelic_rpm gem for context
+            if result.respond_to?(:key)
+              result = result.key
+            elsif result.respond_to?(:data)
+              result = result.data
+            end
+            NewRelic::Agent::Datastores.notice_statement(result, elapsed)
```

Comment:
@StabbyCutyou @alanbrent The only pending issue I see with this PR relates to how to handle when `result` is `nil`. This will be the case when `Bucket#get_or_new` is called and no object yet exists for that key. This is due to the fact that `Client#get_object` ( the wrapped method ) throws a not found exception when no object for the given key exists. However, the `Datastores#wrap` method has an ensure block that executes the callback passing `nil` when a not found exception has occured. 

https://github.com/Tapjoy/rpm/blob/master/lib/new_relic/agent/datastores.rb#L115

Do we want to just add a guard to the `notice_statement` call when result is `nil`? Would this result in not reporting valuable information to new relic?